processor/lsmintervalprocessor/internal/merger/model.go

```diff
@@ -0,0 +1,389 @@
+			mClone.SetEmptySummary()
+		case pmetric.MetricTypeSum:
+			src := m.Sum()
+
+			dest := mClone.SetEmptySum()
+			dest.SetAggregationTemporality(src.AggregationTemporality())
+			dest.SetIsMonotonic(src.IsMonotonic())
+		case pmetric.MetricTypeHistogram:
+			src := m.Histogram()
+
+			dest := mClone.SetEmptyHistogram()
+			dest.SetAggregationTemporality(src.AggregationTemporality())
+		case pmetric.MetricTypeExponentialHistogram:
+			src := m.ExponentialHistogram()
+
+			dest := mClone.SetEmptyExponentialHistogram()
+			dest.SetAggregationTemporality(src.AggregationTemporality())
+		}
+
+		v.metricLookup[metricID] = mClone
+	}
+
+	return mClone, metricID
+}
+
+func merge[DPS DataPointSlice[DP], DP DataPoint[DP]](
+	from, to DPS,
+	mID identity.Metric,
+	lookup map[identity.Stream]DP,
+	temporality pmetric.AggregationTemporality,
+) {
+	switch temporality {
+	case pmetric.AggregationTemporalityCumulative:
+		mergeCumulative(from, to, mID, lookup)
+	case pmetric.AggregationTemporalityDelta:
+		mergeDelta(from, to, mID, lookup)
+	}
+}
+
+func mergeCumulative[DPS DataPointSlice[DP], DP DataPoint[DP]](
+	from, to DPS,
+	mID identity.Metric,
+	lookup map[identity.Stream]DP,
+) {
+	for i := 0; i < from.Len(); i++ {
+		fromDP := from.At(i)
+
+		streamID := identity.OfStream(mID, fromDP)
+		toDP, ok := lookup[streamID]
+		if !ok {
+			toDP = to.AppendEmpty()
+			fromDP.CopyTo(toDP)
+			lookup[streamID] = toDP
+			continue
+		}
+
+		if fromDP.Timestamp() > toDP.Timestamp() {
+			fromDP.CopyTo(toDP)
+		}
+	}
+}
+
+func mergeDelta[DPS DataPointSlice[DP], DP DataPoint[DP]](
+	from, to DPS,
+	mID identity.Metric,
+	lookup map[identity.Stream]DP,
+) {
+	for i := 0; i < from.Len(); i++ {
+		fromDP := from.At(i)
+
+		streamID := identity.OfStream(mID, fromDP)
+		toDP, ok := lookup[streamID]
+		if !ok {
+			toDP = to.AppendEmpty()
+			fromDP.CopyTo(toDP)
+			lookup[streamID] = toDP
+			continue
+		}
+
+		switch fromDP := any(fromDP).(type) {
+		case pmetric.NumberDataPoint:
+			mergeDeltaSumDP(fromDP, any(toDP).(pmetric.NumberDataPoint))
+		case pmetric.HistogramDataPoint:
+			mergeDeltaHistogramDP(fromDP, any(toDP).(pmetric.HistogramDataPoint))
+		}
+
+		// Keep the highest timestamp for the aggregated metric
+		if fromDP.Timestamp() > toDP.Timestamp() {
+			toDP.SetTimestamp(fromDP.Timestamp())
+		}
+	}
+}
+
+func mergeDeltaSumDP(from, to pmetric.NumberDataPoint) {
+	switch from.ValueType() {
+	case pmetric.NumberDataPointValueTypeInt:
+		to.SetIntValue(to.IntValue() + from.IntValue())
+	case pmetric.NumberDataPointValueTypeDouble:
+		to.SetDoubleValue(to.DoubleValue() + from.DoubleValue())
+	}
+}
+
+func mergeDeltaHistogramDP(from, to pmetric.HistogramDataPoint) {
+	// Explicit bounds histogram should have same pre-defined buckets.
+	// However, it is possible that the boundaries got updated. In such
+	// scenarios we can't calculate the histogram for conflicting
+	// boundaries. In practical situations, we should not see such
```

Comment:
Could we assume uniform distribution across buckets, and split the buckets while merging and distribute the value according to the split weight? e.g. if we needed to split the bucket `[1..100]` into `[1..25]` and `[26..100]`, then we would assign 25% of the value to the first bucket and 75% to the second one.

The new buckets would be based on merging overlapping intervals: https://www.enjoyalgorithms.com/blog/merge-overlapping-intervals

(I think this can be a TODO, just saying I don't think it's so intractable)

Reply:
> Could we assume uniform distribution across buckets

I was initially hesitant to make this assumption but now that I think about it, this state will be temporary -- maybe when the service is updating with a new histogram config. And, assuming a temporary state, we can make this assumption.

> The new buckets would be based on merging overlapping intervals:

If we make the assumption of uniform distribution, we can, technically, keep the same buckets as before and normalize the new buckets. Not sure if updating the buckets is a good idea since they are explicit buckets specified by the users. If we change the buckets we might end up with concerns about unknown buckets, though this concern would not be a big deal for our (Elastic's) use cases but it may end up being troublesome if a user just wants to roll up the metrics and index them using this processor.

I will update the TODO and create an issue for this.